common/src/api/external/mod.rs

```diff
@@ -2045,6 +2143,107 @@ impl JsonSchema for L4PortRange {
     }
 }
 
+impl From<L4Port> for L4PortRange {
+    fn from(value: L4Port) -> Self {
+        Self { first: value, last: value }
+    }
+}
+
+impl From<RangeInclusive<L4Port>> for L4PortRange {
+    fn from(value: RangeInclusive<L4Port>) -> Self {
+        let (first, last) = value.into_inner();
+        Self { first, last }
+    }
+}
+
+/// A range of ICMP(v6) types or codes. This range is inclusive on both ends.
+#[derive(
+    Clone, Copy, Debug, DeserializeFromStr, SerializeDisplay, PartialEq,
+)]
+pub struct IcmpParamRange {
+    /// The first number in the range
+    pub first: u8,
+    /// The last number in the range
+    pub last: u8,
+}
+
+impl FromStr for IcmpParamRange {
+    type Err = String;
+    fn from_str(range: &str) -> Result<Self, Self::Err> {
+        const INVALID_NUMBER_MSG: &str = "invalid 8-bit number";
+
+        match range.split_once('-') {
+            None => {
+                let port = range
+                    .parse::<u8>()
+                    .map_err(|_| INVALID_NUMBER_MSG.to_string())?;
+                Ok(IcmpParamRange { first: port, last: port })
+            }
+            Some((left, right)) => {
+                let first = left
+                    .parse::<u8>()
+                    .map_err(|_| INVALID_NUMBER_MSG.to_string())?;
+                let last = right
+                    .parse::<u8>()
+                    .map_err(|_| INVALID_NUMBER_MSG.to_string())?;
```

Comment:
nitpicky: it looks like these strings never include dynamically-formatted content, why not just make the error type `&'static str` so we don't have to `.to_string()` them every time?

certainly not a code path that it actually matters to optimize for not allocating but, :woman_shrugging:

Reply:
Thanks -- I'll fix this (and `L4PortRange`) up.

Reply:
Should be done in 5c7caed.